File: TH1/cknn.cpp
#include <iostream>
#include <vector>
#include <queue>

using namespace std;

const int INF = 1e9;

// Hàm t?m cây khung nh? nh?t b?ng thu?t toán Prim
int Prim(vector<vector<int> >& graph) {
    int n = graph.size();
    vector<bool> visited(n, false); // Ðánh d?u các ð?nh ð? thãm
    priority_queue<pair<int, int>, vector<pair<int, int> >, greater<pair<int, int> > > pq; // Hàng ð?i ýu tiên

    int minCost = 0;

    // B?t ð?u t? ð?nh 0
    pq.push(make_pair(0, 0));

    while (!pq.empty()) {
        int u = pq.top().second;
        int w = pq.top().first;
        pq.pop();

        if (visited[u]) continue;

        visited[u] = true;
        minCost += w;

        for (int v = 0; v < n; ++v) {
            if (!visited[v] && graph[u][v] != INF) {
                pq.push(make_pair(graph[u][v], v));
            }
        }
    }

    return minCost;
}

int main() {
    int n; // S? ð?nh
    cin >> n;

    vector<vector<int> > graph(n, vector<int>(n));

    // Nh?p ma tr?n tr?ng s?
    for (int i = 0; i < n; ++i) {
        for (int j = 0; j < n; ++j) {
            cin >> graph[i][j];
            if (graph[i][j] == 0) {
                // N?u có giá tr? 0, ðánh d?u là vô cùng
                graph[i][j] = INF;
            }
        }
    }

    int minCost = Prim(graph);
    cout << "T?ng tr?ng s? c?a cây khung nh? nh?t: " << minCost << endl;

    return 0;
}

File: TH1/cknn.py
import sys

# thuật toán Prim
def prim(n, graph):
    min_cost = 0    # Biến lưu tổng trọng số của cây khung nhỏ nhất
    visited = [False] * n  
    min_weights = [float('inf')] * n  # Danh sách trọng số nhỏ nhất đến từng đỉnh

    # Bắt đầu từ đỉnh 0
    min_weights[0] = 0

    for _ in range(n):
        u = -1  # Đỉnh chưa thăm có trọng số nhỏ nhất
        for v in range(n):
            if not visited[v] and (u == -1 or min_weights[v] < min_weights[u]):
                u = v
        
        visited[u] = True
        min_cost += min_weights[u]

        # Cập nhật trọng số nhỏ nhất đến các đỉnh kề của u
        for v in range(n):
            if not visited[v] and graph[u][v] < min_weights[v]:
                min_weights[v] = graph[u][v]

    return min_cost

if __name__ == "__main__":
    filename = "data.txt"
    graph = []

    with open(filename, "r") as file:
        n = int(file.readline())
        for line in file:
            row = line.split()
            for j in range(n):
                if row[j] == '0':
                    row[j] = float('inf')
                else:
                    row[j] = int(row[j])
            graph.append(row)

    min_cost = prim(n, graph)  # Tìm cây khung nhỏ nhất
    print("Tổng trọng số của cây khung nhỏ nhất:", min_cost)  # In kết quả lên màn hình
